sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedReader.java

```diff
@@ -214,6 +214,12 @@ public boolean advance() throws IOException {
         curTimestamp =
             pState.timestampPolicy.getTimestampForRecord(pState.mkTimestampPolicyContext(), record);
         curRecord = record;
+        if (this.offsetBasedDeduplicationSupported) {
```

Comment:
instead of doing the encoding here, you could just do it within the currentOffset and currentId methods?

Benefits are if the runner isn't doing offsetbaseddeduplication there wouldn't be overhead, and just basing upon the curRecord means there is less variables to keep in sync.
Downside is more encoding if we call the offset or id methods more than once but I don't think that is the case.

Reply:
Moved encoding, yeah this looks simpler and they should only be called once.